# supervision worksheets for bioinformatics module (part II Computer Science, University of Cambridge) 


https://www.cl.cam.ac.uk/teaching/1718/Bioinfo/

Excerpt from the course webpage 2017:

## Aims

This course focuses on algorithms used in Bioinformatics and System Biology. Most of the algorithms are general and can be applied in other fields on multidimensional and noisy data. All the necessary biological terms and concepts useful for the course and the examination will be given in the lectures. The most important software implementing the described algorithms will be demonstrated.

## Lectures

**Introduction to biological data**: Bioinformatics as an interesting field in computer science.

**Dynamic programming**. Longest common subsequence, DNA global and local alignment, linear space alignment, Nussinov algorithm for RNA, heuristics for multiple alignment. (Vol. 1, chapter 5)

**Sequence database search**. Blast. (see notes and textbooks)

**Genome sequencing**. De Bruijn graph. (Vol. 1, chapter 3)

**Phylogeny**. Distance based algorithms (UPGMA, Neighbour-Joining). Parsimony-based algorithms. (Vol. 2, chapter 7)

**Clustering**. Hard and soft K-means clustering, use of Expectation Maximization in clustering, Hierarchical clustering, Markov clustering algorithm. (Vol. 2, chapter 8)

**Genomics Pattern Matching**. Suffix Tree String Compression and the Burrows-Wheeler Transform. (Vol. 2, chapter 9)

**Hidden Markov Models**. The Viterbi algorithm, profile HMMs for sequence alignment, classifying proteins with profile HMMs, soft decoding problem, Baum-Welch learning. (Vol. 2, chapter 10) 
